scripts/pseudo_hadd.py

```diff
@@ -0,0 +1,144 @@
+        first_source.cd(path)
+        obj = key.ReadObj()
+
+        if isinstance(obj, ROOT.TObjString):
+            for fname in sourcefiles:
+                f = ROOT.TFile.Open(fname, "READ")
+                f.cd(path)
+                key2 = ROOT.gDirectory.GetListOfKeys().FindObject(key.GetName())
+                string = key2.ReadObj().Clone()
+                f.Close()
+                target.cd()
+                string.Write(key.GetName())
+
+        elif isinstance(obj, ROOT.TH1):
+            # descendant of TH1 -> merge it
+            print("[INFO] Merging histogram {}/{}".format(path, obj.GetName()))
+            h1 = obj.Clone()
+            # loop over all source files and add the content of the
+            # corresponding histogram to the one pointed to by "h1"
+            for fname in sourcefiles[1:]:
+                f = ROOT.TFile.Open(fname, "READ")
+                # make sure we are at the correct directory level by cd'ing to path
+                f.cd(path)
+                key2 = ROOT.gDirectory.GetListOfKeys().FindObject(h1.GetName())
+                if key2:
+                    h2 = key.ReadObj()
+                    h1.Add(h2)
+                    del h2
+                f.Close()
+            target.cd()
+            h1.Write(key.GetName())
+        
+        elif isinstance(obj, ROOT.TTree):
+            tree_name = obj.GetName()
+            tchain = ROOT.TChain(tree_name)
+            full_tree_path = "{}/{}".format(path, tree_name)
+            print("[INFO] Linking trees in {} files under '{}'...".format(len(sourcefiles), full_tree_path))
+            for fname in sourcefiles:
+                if not tchain.AddFile(fname, ROOT.TTree.kMaxEntries, full_tree_path):
+                    raise RuntimeError('[ERROR] Could not connect tree {} in file {} to TChain!'.format(full_tree_path, fname))
+                # test the links
+                if check:
+                    num_entries = tchain.GetEntries()
+                    print("[INFO] Linking successful. Combined TChain yielded {} entries.".format(num_entries))
+            target.cd()
+            tchain.Write()
+        
+        elif isinstance(obj, ROOT.TDirectory):
+            # it's a subdirectory
+            print("[INFO] Found subdirectory " + obj.GetName())
+            
+            # create a new subdir of same name and title in the target file
+            target.cd()
+            newdir = target.mkdir(obj.GetName(), obj.GetTitle())
+            
+            # newdir is now the starting point of another round of merging
+            # newdir still knows its depth within the target file via
+            # GetPath(), so we can still figure out where we are in the recursion
+            MergeRootFiles(newdir, sourcefiles, check=check)
+            
+        else:
+            print("[WARNING] Unknown object type, name: " + obj.GetName() + " title: " + obj.GetTitle())
+    target.SaveSelf(ROOT.kTRUE)
+    ROOT.TH1.AddDirectory(status)
```

Comment:
Why do you need to set the directory status of the TH1 object above to `false` above? Is it really necessary?

Reply:
This apparently reduces memory usage: See https://root.cern.ch/doc/master/hadd_8C_source.html Line 65

Reply:
Okay, makes sense, since you don't create references to the objects.